### ISSUE TITLE
[Contributor Welcome] Implement C++ API version of torch.nn.functional.normalize

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#27048]
## Context 
We would like to add `torch::nn::functional::normalize` to the C++ API, so that C++ users can easily find the equivalent of Python API `torch.nn.functional.normalize`.

## Steps
- Add `torch::nn::NormalizeOptions` to `torch/csrc/api/include/torch/nn/options/normalization.h` (add this file if it doesn’t exist), which should include the following parameters (based on https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.normalize)
    - `TORCH_ARG(double, p) = 2.0`
    - `TORCH_ARG(int64_t, dim) = 1`
    - `TORCH_ARG(double, eps) = 1e-12`
    - NOTE: please make sure to add the same comments for the parameters as in the Python version. For example, for parameter `p` we should say `the exponent value in the norm formulation. Default: 2`
- Add `torch::nn::functional::normalize(...)` in  `torch/csrc/api/include/torch/nn/functional/normalization.h` (add this file if it doesn’t exist). The function should have the following signature:

```
namespace torch {
namespace nn {
namespace functional {

inline Tensor normalize(
    const Tensor& input,
    const NormalizeOptions& options = {},
    c10::optional<Tensor> out = c10::nullopt) {
  ...
}

} // namespace functional
} // namespace nn
} // namespace torch
```

- Add test for `torch::nn::functional::normalize(...)` in test/cpp/api/functional.cpp. It can just contain tests for both “out is null” and “out is non-null” cases, and makes sure that for “out is null” case the function is differentiable. (We should get the expected values from the corresponding Python version.)

## Helpful Resources
There are quite a few PRs for adding new functionals / new modules for the C++ API (the list of PRs is in https://github.com/pytorch/pytorch/issues/25883), which can serve as great references. Also please ping @yf225 on this issue if you encounter any problems.

## How do I claim this feature request?
Please comment in this issue if you are interested in working on it.

cc @yf225

**Comment by gtamba** [comment]
Hey I'd like to work on this issue, reading the resource materials now I should have a PR soon if I don't run into problems.

**Comment by gtamba** [comment]
Hey I'd like to work on this issue, reading the resource materials now I should have a PR soon if I don't run into problems.

**Comment by yf225** [comment]
Thanks @gtamba ! Please let me know if you need anything from me :)

**Comment by gtamba** [comment]
@yf225 : I spent longer then I expected studying the codebase so apologies for a delayed update, but I have a few doubts that I guess I can't figure out by myself. 

The code logic to be reproduced in CPP in Python is :

```python
    if out is None:
        denom = input.norm(p, dim, True).clamp_min(eps).expand_as(input)
        ret = input / denom
    else:
        denom = input.norm(p, dim, True).clamp_min(eps).expand_as(input)
        ret = torch.div(input, denom, out=out)
    return ret

```

On the library API reference I could only find function signatures of 

```cpp
at::norm(const Tensor &self, c10::optional<Scalar> p, DimnameList dim, bool keepdim = false);
at::clamp_min(const Tensor &self, Scalar min);
```

But not of `at::expand_as()`  while using the API [search](https://pytorch.org/cppdocs/search.html?q=expand&check_keywords=yes&area=default#) . 

Could you give me any pointers on how I can figure out which cpp code / files are called from the Python interface?

**Comment by yf225** [comment]
@gtamba My sincere apologies for the bad experience - the Tensor API doc is currently broken (https://github.com/pytorch/pytorch/issues/25845) and I will try to fix it as soon as possible. Regarding the equivalent code in C++, since functions such as `clamp_min`/`expand_as` are methods of Tensor, I think the code can look like:
```cpp
    if (out == c10::nullopt) {
        auto denom = input.norm(p, dim, true).clamp_min(eps).expand_as(input);
        return input / denom;  // or `input.div(denom)` if this doesn't work
    } else {
        auto denom = input.norm(p, dim, true).clamp_min(eps).expand_as(input);
        return torch::div_out(*out, input, denom);
    }
```
Thanks so much for your help!

**Comment by gtamba** [comment]
@yf225 : No apologies needed at all I could have tried the same calls implicitly but got a little confused by all the many references to norms norm layers and normalise functions, thank you so much for the code snippet I still have a lot to learn :

I noticed that the Python [test](https://github.com/pytorch/pytorch/blob/f4f6d8dda5bb93145dc78005d1c1c3be9be8f876/test/test_nn.py#L3262) does not check for correctness but only for gradient correctness (assuming the functions it composes from are tested individually), is this sufficient for the C++ test or should I put in arbitrary solutions to check against this as well?

**Comment by gtamba** [comment]
@yf225 : No apologies needed at all I could have tried the same calls implicitly but got a little confused by all the many references to norms norm layers and normalise functions, thank you so much for the code snippet I still have a lot to learn :

I noticed that the Python [test](https://github.com/pytorch/pytorch/blob/f4f6d8dda5bb93145dc78005d1c1c3be9be8f876/test/test_nn.py#L3262) does not check for correctness but only for gradient correctness (assuming the functions it composes from are tested individually), is this sufficient for the C++ test or should I put in arbitrary solutions to check against this as well?

**Comment by yf225** [comment]
@gtamba My apologies that I was out of office last week which causes the delay in response - I saw that you add the tests in `test/cpp/api/functional.cpp` which is the right place to put them, and the tests you added for checking the expected output value are sufficient as well. :D Thanks so much for the great work!
